core/preconditioner/isai.cpp

```diff
@@ -97,39 +98,79 @@ convert_to_csr_and_sort(std::shared_ptr<const Executor> &exec, const LinOp *mtx,
+ */
+template <typename Csr>
+std::shared_ptr<const Csr> extend_sparsity(
+    std::shared_ptr<const Executor> &exec, std::shared_ptr<const Csr> mtx,
+    int power, bool lower)
+{
+    GKO_ASSERT_EQ(power >= 1, true);
+    if (power == 1) {
+        return mtx;
+    }
+    auto id = mtx->clone();
+    exec->run(isai::make_identity_triangle(id.get(), lower));
+    auto id_power = id->clone();
+    auto tmp = Csr::create(exec, mtx->get_size());
+    // compute id^(n-1) and then multiply it with mtx
+    // TODO replace this by a square-and-multiply algorithm
+    for (int i = 1; i < power - 1; ++i) {
+        id->apply(id_power.get(), tmp.get());
+        std::swap(id_power, tmp);
+    }
+    // finally compute id^(n-1) * mtx
+    id_power->apply(mtx.get(), tmp.get());
+    return tmp;
 }
 
 
 template <isai_type IsaiType, typename ValueType, typename IndexType>
 void Isai<IsaiType, ValueType, IndexType>::generate_l_inverse(
-    const LinOp *to_invert_l, bool skip_sorting)
+    std::shared_ptr<const LinOp> to_invert_l, bool skip_sorting, int power)
 {
     GKO_ASSERT_IS_SQUARE_MATRIX(to_invert_l);
     auto exec = this->get_executor();
     auto csr_l = convert_to_csr_and_sort<Csr>(exec, to_invert_l, skip_sorting);
-    const auto num_elems = csr_l->get_num_stored_elements();
+    auto strategy = csr_l->get_strategy();
+    auto csr_extended_l = extend_sparsity(exec, csr_l, power, true);
+    const auto num_elems = csr_extended_l->get_num_stored_elements();
 
     std::shared_ptr<Csr> inverted_l =
-        Csr::create(exec, csr_l->get_size(), num_elems, csr_l->get_strategy());
-    exec->run(isai::make_generate_l_inverse(csr_l.get(), inverted_l.get()));
+        Csr::create(exec, csr_extended_l->get_size(), num_elems, strategy);
+    exec->run(
+        isai::make_generate_l_inverse(csr_extended_l.get(), inverted_l.get()));
 
     approximate_inverse_ = std::move(inverted_l);
 }
 
 
 template <isai_type IsaiType, typename ValueType, typename IndexType>
 void Isai<IsaiType, ValueType, IndexType>::generate_u_inverse(
-    const LinOp *to_invert_u, bool skip_sorting)
+    std::shared_ptr<const LinOp> to_invert_u, bool skip_sorting, int power)
 {
     GKO_ASSERT_IS_SQUARE_MATRIX(to_invert_u);
     auto exec = this->get_executor();
     auto csr_u = convert_to_csr_and_sort<Csr>(exec, to_invert_u, skip_sorting);
-    const auto num_elems = csr_u->get_num_stored_elements();
+    auto strategy = csr_u->get_strategy();
+    auto csr_extended_u = extend_sparsity(exec, csr_u, power, false);
+    const auto num_elems = csr_extended_u->get_num_stored_elements();
 
     std::shared_ptr<Csr> inverted_u =
-        Csr::create(exec, csr_u->get_size(), num_elems, csr_u->get_strategy());
-    exec->run(isai::make_generate_u_inverse(csr_u.get(), inverted_u.get()));
+        Csr::create(exec, csr_extended_u->get_size(), num_elems, strategy);
+    exec->run(
+        isai::make_generate_u_inverse(csr_extended_u.get(), inverted_u.get()));
```

Comment:
What we could do alternatively is to forward the `csr_u` to the `generate_u_inverse`, and use the `csr_extended_u` as the template for the `inverted_u`.
We would need to remove the copying of the sparsity information inside the `generate_x_inverse`, and change the `clone_allocations` in the reference implementation to not override the sparsity pattern.

If you want, I can push these changes directly into this PR. The OMP and CUDA/HIP implementations would also be easy to change to that. PR #502 would then, however, depend on this PR to finish.

Reply:
That is a really great idea, since otherwise we would mostly be reading zeros. Similar to what we do in ParILU vs. ParILUT

Reply:
Now that I think of it, it is probably better to do that at a later date, so we can move forward for now. This is just a small optimization, we could actually do that together with #511